### PR TITLE
fix: Make sure outstanding RPCs count in ChannelPool can not go negative

### PR DESCRIPTION
**Pull request by blakeli0**
fixes: #2081 

Add two flags `wasClosed` and `wasReleased` to `ReleasingClientCall` to check various scenarios. The combination of these two flags can make sure the count of outstanding RPCs can never go negative, and help us identify what exactly goes wrong next time it happens. 

### More background for the purpose of `outstandingRPCs`

The primary purpose of keeping a count for outstanding RPCs is to track when a channel is
safe to close. In grpc, initialization & starting of rpcs is split between 2 methods:
Channel#newCall() and ClientCall#start. gRPC already has a mechanism to safely close channels
that have rpcs that have been started. However, it does not protect calls that have been
created but not started. In the sequence: Channel#newCall() Channel#shutdown()
ClientCall#Start(), gRpc will error out the call telling the caller that the channel is
shutdown.
Hence, the increment of outstanding RPCs has to happen when the ClientCall is initialized,
as part of Channel#newCall(), not after the ClientCall is started. The decrement of
outstanding RPCs has to happen when the ClientCall is closed or the ClientCall failed to
start.

### CODE REVIEW

File: gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java [hunk 1]
```diff
@@ -542,17 +544,32 @@ public void start(Listener<RespT> responseListener, Metadata headers) {
...
             headers);
       } catch (Exception e) {
         // In case start failed, make sure to release
-        entry.release();
+        if (wasReleased.compareAndSet(false, true)) {
+          entry.release();
+        } else {
+          LOG.log(
+              Level.WARNING,
+              "The entry is already released, this indicates that onClose() has already being called previously");
```
Comment (meltsufin) [inline gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java:571]:
```suggestion
              "The entry is already released. This indicates that onClose() has already been called previously");
```

**Review by meltsufin — APPROVED** [review]
I'm not sure if all of these logs should be warnings, but I think the safeguards make sense.

File: gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java [hunk 2]
```diff
@@ -542,17 +560,32 @@ public void start(Listener<RespT> responseListener, Metadata headers) {
             new SimpleForwardingClientCallListener<RespT>(responseListener) {
               @Override
               public void onClose(Status status, Metadata trailers) {
-                try {
+                if (wasClosed.compareAndSet(false, true)) {
                   super.onClose(status, trailers);
-                } finally {
-                  entry.release();
+                  if (wasReleased.compareAndSet(false, true)) {
+                    entry.release();
```
Comment (igorbernstein2) [inline gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java:566]: This needs to be in a finally block so that ref count can be decremented even if the caller's responseListener throws an error. In other words it doesnt matter if the caller's handler has a bug, from the channel perspective, the call is done and the ref count should be decremented

File: gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java [hunk 3]
```diff
@@ -542,17 +560,32 @@ public void start(Listener<RespT> responseListener, Metadata headers) {
...
+                  } else {
+                    LOG.log(
+                        Level.WARNING,
+                        "Call was not closed but entry was released. This may be due to an exception on start of the call.");
```
Comment (igorbernstein2) [inline gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java:570]: I would reword this to "channel refcount was released before the rpc closed. This may be due to ..."

File: gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java [hunk 4]
```diff
@@ -526,6 +542,8 @@ public <RequestT, ResponseT> ClientCall<RequestT, ResponseT> newCall(
   static class ReleasingClientCall<ReqT, RespT> extends SimpleForwardingClientCall<ReqT, RespT> {
     @Nullable private CancellationException cancellationException;
     final Entry entry;
+    private final AtomicBoolean wasClosed = new AtomicBoolean();
+    private final AtomicBoolean wasReleased = new AtomicBoolean();
```
Comment (igorbernstein2) [inline gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java:546]: Why both bools?

Comment (blakeli0) [reply gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java:546]: So that we can identify different scenarios next time it happens
1. Multiple calls to onClose()
2. onClose() being called first, try-catch of `start()` being called second
3. try-catch of `start()` being called first, onClose() being called second.

With only one flag, I don't think we can identify these different scenarios.

File: gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java (hunk 3 above)
Comment (blakeli0) [reply gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java:570]: Because `channel refcount` is easier to understand than `entry`? Or put more focus on `released before closed`?

File: gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java (hunk 2 above)
Comment (blakeli0) [reply gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java]: Updated

File: gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java [hunk 5]
```diff
@@ -542,17 +560,35 @@ public void start(Listener<RespT> responseListener, Metadata headers) {
...
+                      LOG.log(
+                          Level.WARNING,
+                          "Entry was released before the call is closed. This may be due to an exception on start of the call.");
+                    }
+                  }
+                } else {
+                  LOG.log(
+                      Level.WARNING,
+                      "Call is being closed more than once. Please make sure that onClose() is not being manually called.");
                 }
```
Comment (igorbernstein2) [inline gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java:579]: The nesting makes it hard to read, I think early returns would make it easier:

```java
if (wasClosed.compareAndSet(false, true)) {
   LOG.log(
                      Level.WARNING,
                      "Call is being closed more than once. Please make sure that onClose() is not being manually called.");
   return;
}

if (!wasReleased.compareAndSet(false, true)) {
    LOG.log(
                      Level.WARNING,
                      "Call is being closed more than once. Please make sure that onClose() is not being manually called.");
    return;
}
entry.release();
```

Comment (blakeli0) [reply gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java]: Thanks for the suggestion! Yeah with try-finally being added, it's getting harder to read. I extracted the checking of `wasClosed` to earlier to remove one indent. Kept the checking of `wasRelease` in finally because I think it's logically easier to understand even though has more indents.
